### PR TITLE
SDI-941 Create all adjudications and delete by migration id

### PR DESCRIPTION
**Pull request by andymarke**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/AdjudicationsMappingResource.kt [hunk 1]
```diff
@@ -73,6 +74,48 @@ class AdjudicationsMappingResource(private val mappingService: AdjudicationMappi
       )
     }
 
+  @PreAuthorize("hasRole('ROLE_NOMIS_ADJUDICATIONS')")
+  @PostMapping("/mapping/adjudications/all")
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/AdjudicationsMappingResource.kt:78]: have not added individual punishment and hearings yet. But will be required for synchronisation

File: src/main/resources/db/migration/V1_28__adjudication_hearings.sql [hunk 2]
```diff
@@ -0,0 +1,11 @@
+create table adjudication_hearing_mapping
+(
+    nomis_hearing_id bigint                   not null,
+    dps_hearing_id   varchar(30)              not null primary key,
+    when_created     timestamp with time zone not null default now(),
+    label            varchar(20),
+    mapping_type     varchar(20)              not null
+);
+create index adjudication_hearing_mapping_nomis_id_index on adjudication_hearing_mapping (nomis_hearing_id);
```
Comment (andymarke) [inline src/main/resources/db/migration/V1_28__adjudication_hearings.sql:9]: nomis_hearing_id is not unique since this is a one to many

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationHearingMappingDto.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  @field:Size(max = 20)
+  val label: String? = null,
+
+  @Schema(
+    description = "Mapping type",
+    allowableValues = ["MIGRATED", "ADJUDICATION_CREATED"],
+    defaultValue = "ADJUDICATION_CREATED",
+  )
+  @field:Size(max = 20)
+  val mappingType: String? = null,
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationHearingMappingDto.kt:29]: Is that defaultValue above related the default here?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationPunishmentMappingDto.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+@Schema(description = "Adjudication punishment (aka award) mapping")
+data class AdjudicationPunishmentMappingDto(
+
+  @Schema(description = "NOMIS booking id", required = true, example = "123456")
+  val nomisBookingId: Long,
+
+  @Schema(description = "NOMIS sanction sequence", required = true, example = "4")
+  val nomisSanctionSequence: Int,
+
+  @Schema(description = "DPS hearing id", required = true, example = "123456")
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationPunishmentMappingDto.kt:19]: Not hearing id - punishment id?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationHearingMappingDto.kt (hunk 3 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationHearingMappingDto.kt:29]: Yes and no. There are 2 ways to create this mapping, via migration using the DTO above so this is nested. That will always be MIGRATED, or on it own (the endpoint hasn't been written yet) that will default to "ADJUDICATION_CREATED"

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationPunishmentMappingDto.kt (hunk 4 above)
Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/data/AdjudicationPunishmentMappingDto.kt:19]: yes, I'll submit a new PR after this is merged

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/AdjudicationMappingService.kt [hunk 5]
```diff
@@ -11,21 +11,56 @@ import org.springframework.data.domain.Pageable
...
   private val telemetryClient: TelemetryClient,
 ) {
   private companion object {
     private val log = LoggerFactory.getLogger(this::class.java)
   }
 
+  @Transactional
+  suspend fun createMapping(createMappingRequest: AdjudicationAllMappingDto) {
+    createMapping(
+      createMappingRequest.adjudicationId.copy(
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/AdjudicationMappingService.kt:41]: Why the use of `copy` here?

Comment (andymarke) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/AdjudicationMappingService.kt:41]: because we always want the mappingType and label to equal the parent mapping.
This is a special case where a migration can create *many* mapping records and they all need to either fail or succeed

File: src/test/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/AdjudicationMappingResourceIntTest.kt [hunk 6]
```diff
@@ -291,6 +342,260 @@ class AdjudicationMappingResourceIntTest : IntegrationTestBase() {
...
+        .exchange()
+        .expectStatus().isCreated
+
+      val createdMappingByNomisId = webTestClient.get()
+        .uri("/mapping/adjudications/adjudication-number/$ADJUDICATION_NUMBER/charge-sequence/$CHARGE_SEQ")
+        .headers(setAuthorisation(roles = listOf("ROLE_NOMIS_ADJUDICATIONS")))
+        .exchange()
+        .expectStatus().isOk
+        .expectBody(AdjudicationMappingDto::class.java)
+        .returnResult().responseBody!!
```
Comment (mikehalmamoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/AdjudicationMappingResourceIntTest.kt:405]: It's unusual for you not to use `jsonPath` - any reason?

Comment (andymarke) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/resource/AdjudicationMappingResourceIntTest.kt:405]: probably a cut an paste of the original test. So no "good" reason

**Review by mikehalmamoj — APPROVED** [review]
